File: spl-net/src/main/java/bgu/spl171/net/impl/TFTP/EncoderDecoders/TFTPEncoderDecoder.java
package bgu.spl171.net.impl.TFTP.EncoderDecoders;

import bgu.spl171.net.api.MessageEncoderDecoder;
import bgu.spl171.net.impl.TFTP.CommandBuilders.CommandBuilder;
import bgu.spl171.net.impl.TFTP.CommandEncoders.CommandEncoder;
import bgu.spl171.net.impl.TFTP.Commands.Command;

import java.nio.ByteBuffer;

public class TFTPEncoderDecoder implements MessageEncoderDecoder<Command> {

    ShortEncoderDecoder shortEncoderDecoder = new ShortEncoderDecoder();
    CommandBuilder builder = null;

    @Override
    public Command decodeNextByte(byte nextByte) {
        Command returnValue = null;

        if (builder == null) {
            Short opCode = shortEncoderDecoder.decodeNextByte(nextByte);
            if (opCode != null) {
                builder = CommandBuilder.createBuilder(opCode);
                if (builder == null) {
                    returnValue = CommandBuilder.createCommand(opCode);
                }
            }
        } else {
            returnValue = builder.decodeNextByte(nextByte);
            if (returnValue != null) {
                builder = null;
            }
        }
        return returnValue;
    }

    @Override
    public byte[] encode(Command message) {
        return CommandEncoder.get(message.getOpCode()).encode(message);
    }

}

File: java to c++ encdec/TFTPEncoderDecoder.cpp
#include "TFTPEncoderDecoder.h"
#include "ShortEncoderDecoder.h"

namespace bgu
{
	namespace spl171
	{
		namespace net
		{
			namespace impl
			{
				namespace TFTP
				{
					namespace EncoderDecoders
					{
						using bgu::spl171::net::api::MessageEncoderDecoder;
						using bgu::spl171::net::impl::TFTP::CommandBuilders::CommandBuilder;
						using bgu::spl171::net::impl::TFTP::CommandEncoders::CommandEncoder;
						using bgu::spl171::net::impl::TFTP::Commands::Command;

						Command *TFTPEncoderDecoder::decodeNextByte(char nextByte)
						{
							Command *returnValue = nullptr;

							if (builder == nullptr)
							{
								boost::optional<short> opCode = shortEncoderDecoder->decodeNextByte(nextByte);
								if (opCode)
								{
									builder = CommandBuilder::createBuilder(opCode);
									if (builder == nullptr)
									{
										returnValue = CommandBuilder::createCommand(opCode);
									}
								}
							}
							else
							{
								returnValue = builder->decodeNextByte(nextByte);
								if (returnValue != nullptr)
								{
									builder = nullptr;
								}
							}
							return returnValue;
						}

						std::vector<char> TFTPEncoderDecoder::encode(Command *message)
						{
							return CommandEncoder->get(message->getOpCode()).encode(message);
						}
					}
				}
			}
		}
	}
}
